Removed undocumented mid-p correction to p-values in exact test of Hardy-Weinberg equilibrium and updated corresponding tests.

samuelklee opened the pull request:
Work is split into two commits:

- Removed undocumented mid-p correction to p-values in exact test of Hardy-Weinberg equilibrium and updated corresponding unit tests.
- Updated expected ExcessHet values in integration test resources and added an update toggle to GnarlyGenotyperIntegrationTest.

Various scout cleanups as well.

We now report the same value as ExcHet in bcftools. Note that previous values of 3.0103 (corresponding to mid-p values of 0.5) will now be 0.0000. See discussion below and in linked issue for additional details.

Closes #7392.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -137,11 +137,6 @@ static double exactTest(final int hetCount, final int refCount, final int homCou
         final int rareCopies = 2 * obsHomR + hetCount;
         final int N = hetCount + obsHomC + obsHomR;
 
-        //If the probability distribution has only 1 point, then the mid p-value is .5
```
Note that this short circuit for calculating a mid p-value is incorrect, since we are calculating one-sided p-values here. I just went ahead and removed the short circuit.

samuelklee replied:
See discussion below about the mid-p correction, but I think it's probably fine to remove the short circuit anyway.

samuelklee commented:
Oof, looks like there are now a bunch of broken integration tests that check ExcessHet for whatever reason. So let's definitely decide on whether we want to make the switch to mid p-values before I go through those.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,283 @@
     @DataProvider(name = "smallSets")
     public Object[][] counts() {
         return new Object[][]{
-                {1, 0, 0, 0.5},
```
Consequential test changes start here; most of the other changes in the class are just the result of scout-level cleanup.

samuelklee commented:
Also just realized there’s *yet another* implementation in htsjdk, HardyWeinbergCalculation, so just a reminder to myself to check against that. Looks like a two-sided p-value of sorts is calculated there, but I think it might be non-canonical in that it doesn’t sum only over het counts with likelihoods less than or equal to that of the observed count—it instead sums over all het counts other than that observe. Need to take a closer look.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -193,20 +190,19 @@ static double exactTest(final int hetCount, final int refCount, final int homCou
             return rightPval;
         }
-        return rightPval + StatUtils.sum(Arrays.copyOfRange(probs, hetCount + 1, probs.length)) / mysum;
+        return Math.min(1., rightPval + StatUtils.sum(Arrays.copyOfRange(probs, hetCount + 1, probs.length)) / mysum);
```
BTW, seems like we should use something from QualityUtils for converting this to Phred scale in the calculateEH method, but will let the reviewer decide. Not sure if we will want to e.g. introduce any caps.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,286 @@
+                {1, 1, 0, 1.},
+                // The following test cases are taken from Table 1 of Wigginton et al.
+                // at https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1199378/.
+                // Results were calculated using the python implementation at
```
Another thought: I could put a Java implementation of this slower explicit calculation here in the test code. We could check this against the test cases provided in the paper, and then use it to generate expected results for the additional test cases (rather than hardcoding the results of running the python implementation). Might save future developers a bit of time, although I'm not sure how bulletproof such a Java implementation would be...I'm inclined to trust the scipy/numpy implementations of special functions a little more.

ldgauthier replied:
I don't expect to have to modify this code in the future, so don't break your back bending over backwards for future developers.  I think using a public python library in a separate file is great.

samuelklee replied:
For my part, at least, I have been thinking about parameterizing the annotation to allow the exact test to be performed when an estimate of inbreeding/outcrossing rate is available; it's mentioned near the end of Wigginton et al. as an easy modification to the standard recurrence relation. So I might revisit this code shortly.

But thanks---certainly, the less non-trivial test code I have to write for future developers, the better!

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,292 @@
-    private Allele C = Allele.create("C");
-    private int[] hetPLs = {240, 0, 240};
-    private int[] homRefPLs= {0, 60, 600};
+    private static final double DELTA_PRECISION = 1E-6;
```
Most of the following changes are inconsequential cleanup, but note we do change the precision criteria here.

ldgauthier replied:
We require more precision now, yes?

samuelklee commented:
Also added a little something to prevent output of negative-zero scores (since all of those 3.0103s actually became -0.0000s before this fix), see e.g. https://gatk.broadinstitute.org/hc/en-us/community/posts/360056519272-ExcessHet-value-is-0-0000. Not sure if this is something that should be done at the QualityUtils level either.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/gnarlyGenotyper/GnarlyGenotyperEngine.java:
```diff
@@ -225,7 +225,7 @@ else if (variant.hasAttribute(GATKVCFConstants.AS_RAW_QUAL_APPROX_KEY)) {
             final int refCount = Math.max(numCalledAlleles / 2 - gtCounts.get(1) - gtCounts.get(2), 0);
             //homRefs don't get counted properly because ref blocks aren't annotated
             gtCounts.set(0, refCount);
-            final Pair<Integer, Double> eh = ExcessHet.calculateEH(variant, new GenotypeCounts(gtCounts.get(0), gtCounts.get(1), gtCounts.get(2)), numCalledAlleles / 2);
```
Not sure why this unused argument was here, but I took it out.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -82,25 +82,25 @@ public ExcessHet() {
     }
 
     @VisibleForTesting
-    public static Pair<Integer, Double> calculateEH(final VariantContext vc, final GenotypeCounts t, final int sampleCount) {
```
Not sure why this unused argument was here, but I took it out.

ldgauthier replied:
I may have done a sloppy refactor -- it looks like I'm the one who last changed the method signature.  I'm not worried about this, but thanks for the cleanup.

samuelklee commented:
Now that I've spent a day surveying the landscape of p-values, I'm thinking what must've happened is that an undocumented mid-p-value correction was added to the one-tailed p-value we calculate. This is probably a fine thing to calculate, but such a correction is not discussed in the Wigginton paper cited in the annotation docs.

So we could keep the original calculation, while further clarifying what exactly we are calculating in the docs (and add references, if appropriate). In this case, we can just keep the cleanup and improved tests as a bonus. (As a further bonus, we won't lose our old friend 3.0103!)

Or we could move to this calculation. I'm not sure if the original intent was to be more/less conservative in retaining sites during hard filtering prior to VQSR. But since the filtering threshold used in Best Practices seems very conservative, I would guess that we wanted to err on the side of not throwing out possible variants, even if they are pretty out of HWE. Which makes the choice of a mid p-value puzzling, since it's strictly smaller and will thus lead to more rejections (I think, if I've got everything the right way around!)

Happy to go either direction. In the end, I have just reaffirmed my dislike for p-values, which I had hitherto thought to be saturated.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:
```diff
@@ -28,8 +28,23 @@
+    public void assertThatExpectedOutputUpdateToggleIsDisabled() {
+        Assert.assertFalse(UPDATE_EXACT_MATCH_EXPECTED_OUTPUTS, "The toggle to update expected outputs should not be left enabled");
+    }
+
```
I noted it over in #7122 post merge, but there are a few tests below that were commented out. Not sure if there is a plan to restore these, but I did not update the corresponding ExcessHet values.

ldgauthier replied:
The tests pass on master if I add them back in. I think I just forgot to put them back.  (I don't have an efficient and elegant method for running the debugger on the Nth test of a data provider if N > 0 so I just commented out 0 through N-1.)  You have my permission to uncomment and update the exact match again.

samuelklee replied:
Done.

samuelklee commented:
@ldgauthier @meganshand this is ready for review. Please feel free to un/reassign as necessary. @fleharty @vruano if you have some bandwidth, please take a look.

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:
```diff
@@ -28,8 +28,23 @@
+    // instead of actually running the tests. Can be used with "./gradlew test -Dtest.single=GnarlyGenotyperIntegrationTest"
+    // to update all of the exact-match tests at once. After you do this, you should look at the
+    // diffs in the new expected outputs in git to confirm that they are consistent with expectations.
+    public static final boolean UPDATE_EXACT_MATCH_EXPECTED_OUTPUTS = false;
```
Can you add a test that checks if this is false?  That's what we do for the HaplotypeCaller tests as a failsafe.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -111,6 +124,9 @@ public ExcessHet() {
      * be more computationally expensive to calculate accuracy beyond a given threshold. Here we have enough accuracy
      * to filter anything below a p-value of 10E-6.
      *
+     * Discussion at https://github.com/broadinstitute/gatk/issues/7392 and
```
Thanks!

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -218,8 +228,7 @@ public String getRawKeyName() {
      */
-    //@Override
-    public Map<String, Object> annotateRawData(ReferenceContext ref, VariantContext vc, AlleleLikelihoods<GATKRead, Allele> likelihoods) {
+    public static Map<String, Object> annotateRawData(ReferenceContext ref, VariantContext vc, AlleleLikelihoods<GATKRead, Allele> likelihoods) {
```
I'm not sure why this doesn't already implement `ReducibleAnnotation`, but that's why there used to be an `@Override`.  Can you check to see what happens if you add that interface?  It's not unlikely there's a reason this is the exception, but I'd like to remember what it is and document it.

samuelklee replied:
Ah, thanks for that clarification! Hmm, looks like it's missing `getPrimaryRawKey`, `getSecondaryRawKeys`, and `hasSecondaryRawKeys` methods. Not sure what exactly these are supposed to return, but happy to add them if you can enlighten me!

samuelklee replied:
OK, I've reverted these changes and filed https://github.com/broadinstitute/gatk/issues/7564.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:
```diff
@@ -31,7 +31,20 @@
- * <a href='http://www.sciencedirect.com/science/article/pii/S0002929707607356?via%3Dihub'>Wigginton JE, Cutler DJ, Abecasis GR. <i>A Note on Exact Tests of Hardy-Weinberg Equilibrium. American Journal of Human Genetics</i>. 2005;76(5):887-893</a>.
+ * <a href='https://doi.org/10.1086/429864'>Wigginton, Cutler, and Abecasis 2005.</a>
+ * to calculate a right-sided p-value (the probability of getting the observed number of heterozygotes or higher).
+ * <p>However, note that in releases prior to and including GATK 4.2.1.0, an additional mid-p correction
```
There may be a 4.2.2.1 as well...

samuelklee replied:
OK, thanks for the heads up---happy to coordinate with @droazen.

samuelklee replied:
Changed to 4.2.2.0 for now. @droazen let's get this in for 4.2.2.1? Would rather not have to change it again (or deal with any more rebased exact match updates, for that matter!)

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,311 @@
 
 
-    private Genotype makeG(String sample, Allele a1, Allele a2, int... pls) {
+    private static Genotype makeG(final String sample, final Allele a1, final Allele a2, final int... pls) {
```
These things are like dandelions -- you take one out and another two pop up.  This is duplicated code that's also in VariantContextTestUtils (and a bunch of other places), but if you feel like you've already done your boy scout duty you can leave it.

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,311 @@
+        Assert.assertEquals(annots.keySet(), Collections.singleton(GATKVCFConstants.EXCESS_HET_KEY), "annots");
+        Assert.assertEquals(annots.values().size(), 1, "size");
+        Assert.assertEquals(Double.parseDouble((String) annots.values().iterator().next()), expected, DELTA_PRECISION_PARSE, "het");
+        Assert.assertEquals((String) annots.values().iterator().next(), "0.0000", "zero"); // guards against -0.0000 being returned
```
I didn't think about that, so I'm extra glad you're testing it!

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:
```diff
@@ -1,253 +1,311 @@
+    public void smallSets(final int hetCount, final int homrefCount, final int homvarCount, final double expected) {
         final double actual = ExcessHet.exactTest(hetCount, homrefCount, homvarCount);
         Assert.assertEquals(actual, expected, DELTA_PRECISION, "Pass");
+        // we also test swapping homvarCount <-> homrefCount, which should be symmetric
```
Nice!

ldgauthier reviewed (changes requested):
I'm running the warp plumbing tests now (50 exomes incl. NA12878) to see what changes.  The only possible consequence would be changing the filtering threshold in WARP: https://github.com/broadinstitute/warp/blob/62e06b060b11e246c2623bd68351b2d15e3808bd/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl#L51.  Is there any theoretical reason we should change it now that mid-p is gone?

ldgauthier commented:
Now that I think more about it, I'm surprised none of the GenotypeGVCFs tests changed.  I guess they must all be too few samples.

Mostly for my own reference, WARP tests are running here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11738/

ldgauthier commented:
GenotypeGVCFs results look great (no differences in metrics), but I borked the gnarly docker image name.  Gnarly tests running here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11755/

ldgauthier commented:
We do see a slight difference in the results, but it's an improvement!(except two indels that are probably in the noise)

## GenotypeGVCFs
### Test:
type	TP	FP	FN	RECALL	PRECISION
SNP	17441	85	846	0.954	0.995
INDEL	304	91	99	0.754	0.77

### Truth:
type	TP	FP	FN	RECALL	PRECISION
SNP	17431	89	856	0.953	0.995
INDEL	305	92	98	0.757	0.768

## Gnarly
Test:
type	TP	FP	FN	RECALL	PRECISION
SNP	17514	137	773	0.958	0.992
INDEL	304	92	99	0.754	0.768

### Truth:
type	TP	FP	FN	RECALL	PRECISION
SNP	17509	152	778	0.957	0.991
INDEL	305	93	98	0.757	0.766

That's good enough for this squirrel.  :shipit:

samuelklee commented:
Thanks for the review and running those tests, @ldgauthier! Will restore the aforementioned tests and update a few other exact matches in the rebase this afternoon.

You also asked above if there was a theoretical reason to change the threshold. Since it seems the original was relatively arbitrary (at least from what I've been told, happy to be corrected), I think we can leave it. The new annotation is strictly larger, so we will then be slightly more conservative about keeping sites if we leave the threshold fixed. You can think of this as a slight change in the decision boundary in genotype-count space---perhaps I can add some plots to this thread this afternoon to demonstrate.

In practice, what we care about is whether: 1) many sites flicker across the change in boundary after hard filtering, and/or 2) these sites result in discrepancies post-VQSR. I think the tests you ran suggest that we don't need to worry much about the second issue, and I can take a closer look later to check about the first (which will depend simply on the number of samples and the allele frequency spectrum). We can also take a basic look at how things might change with e.g. more samples using the aforementioned plots.

samuelklee commented:
Ah, now that I'm going through and re-updating test resources for a rebase, I see that I missed responding to this:

> Now that I think more about it, I'm surprised none of the GenotypeGVCFs tests changed. I guess they must all be too few samples.

I think some of them did? E.g. I'm having to re-update `src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/gvcf.basepairResolution.includeNonVariantSites.vcf`.

samuelklee commented:
OK, thanks @ldgauthier, I think I've addressed all the comments but one. A little TODO list for my benefit:

- [x] Updated the GATK version in the ExcessHet documentation to 4.2.2.0, but we'll see if I need to revisit that.
- [ ] Not quite sure about the `ReducibleAnnotation` business. Let me know how to make these changes, or else happy to punt and file an issue.
- [ ] I'm not sure I've quite parsed the results of the Jenkins tests, at least in terms of comparing how many sites get hard filtered with/out the change. Where should I be looking at to see the baseline result for that step? Also looks like a lot of results for https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11755/ were call-cached, is that to be expected? Haven't looked at these tests before, so maybe you can walk me through them at some point. But I guess we can be sure that the overall results don't change too much (at least for 50 samples), which is a good start.
- [ ] Didn't quite get to making those plots of the change in decision boundary, will do that tomorrow or later this week.

ldgauthier commented:
Based on gs://broad-gotc-test-results/staging/joint_genotyping/exome/scientific/2021-09-03-11-25-15/gather_vcfs_low_memory/small_callset_high_threshold.vcf.gz (from the console output) there are slightly fewer variants filtered with ExcessHet now, which is expected since you said it was an across-the-board shift.  Expected (old) has 4335 and actual (new) has 4133 -- no new things, just some now pass.  If you can calculate a new equivalent threshold I'd rather use that, but otherwise I'm not overly concerned about the changes.

I'm not concerned about the Jenkins call caching unless it's for the GenotypeGVCFs task where ExcessHet actually gets calculated.

For the ReducibleAnnotation comments, if you just revert your changes (statics, visibility, etc.) and open an issue I'm fine with that.  Admittedly this could be another target for refactoring.

samuelklee commented:
Thanks for those numbers! I'm not sure we can shift the threshold in a uniquely defined way, since things are a function of the allele-frequency spectrum and the number of (non-missing) samples. So definitely glad to hear the impact is minimal and everything behaves as expected.

I'll revert the changes tomorrow and will merge after your final thumbs up, thanks!

samuelklee commented:
Seems like there weren't any exact-match tests to update after the latest release, so this should be ready pending your approval, @ldgauthier!

ldgauthier reviewed (approved):
I've held this up long enough, but it might be worth asking @gbrandt6 if she wants to take a look at the ExcessHet documentation changes in that class.

samuelklee commented:
Thanks @ldgauthier. @gbrandt6 I’d appreciate it if you want to take a look, but I might ask if you can do it by Friday afternoon—I’m out after then through all of next week. Would like to merge before I head out to avoid any more rebasing and/or updating of exact-match tests. Happy to look at any changes to the docs you might make in a subsequent PR, though!

samuelklee commented:
Actually going to go ahead and merge. Got my booster earlier today and might be knocked out tomorrow. Again, happy to look at subsequent doc changes.

gbrandt6 commented:
@samuelklee I didn't get a chance to take a look, I'll see if we need to make any doc changes next week and let you know
